src/metabase/api/collection.clj

```diff
@@ -123,33 +142,37 @@
 
   The here and below keys indicate the types of items at this particular level of the tree (here) and in its
   subtree (below)."
-  [exclude-archived exclude-other-user-collections namespace]
+  [exclude-archived exclude-other-user-collections namespace shallow location]
   {exclude-archived               [:maybe :boolean]
    exclude-other-user-collections [:maybe :boolean]
-   namespace                      [:maybe ms/NonBlankString]}
+   namespace                      [:maybe ms/NonBlankString]
+   shallow                        [:maybe :boolean]
+   location                       [:maybe ms/NonBlankString]}
   (let [exclude-archived? exclude-archived
         exclude-other-user-collections? exclude-other-user-collections
-        coll-type-ids (reduce (fn [acc {:keys [collection_id dataset] :as _x}]
-                                (update acc (if dataset :dataset :card) conj collection_id))
-                              {:dataset #{}
-                               :card    #{}}
-                              (mdb.query/reducible-query {:select-distinct [:collection_id :dataset]
-                                                          :from            [:report_card]
-                                                          :where           [:= :archived false]}))
         colls (cond->>
-                (t2/select Collection
-                  {:where [:and
-                           (when exclude-archived?
-                             [:= :archived false])
-                           (perms/audit-namespace-clause :namespace namespace)
-                           (collection/visible-collection-ids->honeysql-filter-clause
-                            :id
-                            (collection/permissions-set->visible-collection-ids @api/*current-user-permissions-set*))]})
+               (t2/select Collection
+                          {:where [:and
+                                   (when exclude-archived?
+                                     [:= :archived false])
+                                   (perms/audit-namespace-clause :namespace namespace)
+                                   (collection/visible-collection-ids->honeysql-filter-clause
+                                    :id
+                                    (collection/permissions-set->visible-collection-ids @api/*current-user-permissions-set*))]})
```

Comment:
This is still reading all collections into memory, right? And then subsequently filtering it based on the API keys? That seems like it could still be a bottleneck if an instance has a lot of collections—the FE would be calling this endpoint many times, and each time the backend would be reading _every_ collection. That's a lot more DB load than the status quo, where we only read them all once per page load.

Could we instead change the query to only select the collections it needs to return? Could probably just use a `LIKE` operator on the `location` field to select it by prefix, right?